File: code_samples/section6/problems/problem6_4/problem6_4.js
/*
    This file merges K sorted singly linked lists into one sorted list using
    a custom min-heap of ListNode objects (ordered by node.val).

    High-level algorithm (k-way merge):
      1) Push the head of each non-empty list into a MIN-HEAP.
      2) Repeatedly:
           - Pop the smallest node from the heap
           - Append it to the merged result
           - If that node has a next node, push node.next into the heap
      3) Stop when the heap is empty.

    Why a min-heap?
      - At any moment, the next smallest value must be among the current heads
        of the remaining lists.
      - A min-heap allows extracting that smallest head in O(log k).

    Complexity:
      - Let N be the total number of nodes across all lists.
      - Heap size is at most k (one "front" node per list).
      - Each node is pushed/popped at most once → O(N log k) time, O(k) heap space.

    Memory behavior:
      - mergeKLists REUSES existing nodes (it relinks them); no new nodes are created
        except the dummy node used to simplify list construction.
*/

// Basic singly-linked list node.
class ListNode {
    constructor(val, next = null) {
        // Node value (used for ordering)
        this.val = val;

        // Pointer to next node (or null if end of list)
        this.next = next;
    }
}

// Minimal min-heap of ListNode by node.val
class NodeMinHeap {
    constructor() {
        // Backing array representing the binary heap structure.
        // For a min-heap, data[0] is always the smallest node.
        this.data = [];
    }

    /*
        _bubbleUp(i)
        ------------
        Fixes heap after inserting a new element at index i.

        Insertion strategy:
          - push() appends new element at end of array
          - then bubble it up while it is smaller than its parent

        Stops when:
          - it reaches the root, or
          - parent value <= child value (min-heap property satisfied)

        Complexity: O(log n)
    */
    _bubbleUp(i) {
        while (i > 0) {
            const p = Math.floor((i - 1) / 2); // parent index

            // If child >= parent, min-heap property holds.
            if (this.data[i].val >= this.data[p].val) break;

            // Swap child and parent, then continue upward.
            [this.data[i], this.data[p]] = [this.data[p], this.data[i]];
            i = p;
        }
    }

    /*
        _heapify(i)
        -----------
        Fixes heap after removing the root / moving last element to the root.

        Pop strategy:
          - Save root
          - Move last element to root
          - Sift down: swap with smallest child until heap property holds

        Complexity: O(log n)
    */
    _heapify(i) {
        const n = this.data.length;

        while (true) {
            let smallest = i;
            const left = 2 * i + 1;
            const right = 2 * i + 2;

            // Pick the smallest among i and its children (if they exist).
            if (left < n && this.data[left].val < this.data[smallest].val) smallest = left;
            if (right < n && this.data[right].val < this.data[smallest].val) smallest = right;

            // If i is already smallest, heap property holds.
            if (smallest === i) break;

            // Otherwise swap downward and continue.
            [this.data[i], this.data[smallest]] = [this.data[smallest], this.data[i]];
            i = smallest;
        }
    }

    /*
        push(node)
        ----------
        Inserts a ListNode into the heap.

        NOTE:
          - Caller should not pass null, because we access node.val.

        Complexity: O(log n)
    */
    push(node) {
        this.data.push(node);
        this._bubbleUp(this.data.length - 1);
    }

    /*
        pop()
        -----
        Removes and returns the smallest node (root) from the heap.

        Returns:
          - null if heap is empty
          - otherwise the smallest ListNode

        Complexity: O(log n)
    */
    pop() {
        if (!this.data.length) return null;

        if (this.data.length === 1) {
            // Single element is both root and last element.
            return this.data.pop();
        }

        const root = this.data[0];

        // Move last element to root and restore heap property.
        this.data[0] = this.data.pop();
        this._heapify(0);

        return root;
    }

    /*
        size getter
        -----------
        Exposes current heap size.
    */
    get size() {
        return this.data.length;
    }
}

/*
    mergeKLists
    -----------
    Merges an array of list heads into a single sorted list.

    Parameters:
      - lists: array of ListNode heads (may include nulls; may be empty array)

    Returns:
      - head of merged list, or null if there are no nodes to merge

    Implementation details:
      - Uses NodeMinHeap to always choose the smallest available "front" node.
      - Uses a dummy node to simplify append logic.
*/
function mergeKLists(lists) {
    const heap = new NodeMinHeap();

    // Seed heap with the head of each non-empty list.
    for (const node of lists) {
        if (node) heap.push(node);
    }

    // Dummy node lets us build the result without special-casing the first append.
    const dummy = new ListNode(0);
    let cur = dummy;

    // Repeatedly extract the smallest node and push its successor.
    while (heap.size > 0) {
        const node = heap.pop(); // smallest node by value

        // node should never be null here because heap.size > 0,
        // but we keep logic straightforward.
        cur.next = node;
        cur = cur.next;

        // Push the next node from that list, if any.
        if (node.next) heap.push(node.next);
    }

    // Ensure result terminates cleanly (usually already true).
    cur.next = null;

    return dummy.next;
}

// ===== Helpers =====

/*
    buildList
    ---------
    Builds a linked list from an array of numbers.

    Returns:
      - head node, or null if arr is empty

    Note:
      - Allocates new nodes via 'new ListNode(...)'.
*/
function buildList(arr) {
    let head = null;
    let tail = null;

    for (const v of arr) {
        const node = new ListNode(v);

        if (!head) {
            // First node initializes head and tail.
            head = tail = node;
        } else {
            // Append node and advance tail.
            tail.next = node;
            tail = node;
        }
    }

    return head;
}

/*
    printList
    ---------
    Prints a list as space-separated values preceded by a label.

    If head is null, prints only the label and newline.
*/
function printList(label, head) {
    let s = label;
    let cur = head;

    while (cur) {
        s += cur.val + " ";
        cur = cur.next;
    }

    console.log(s);
}

// ===== Test Cases =====

/*
    Tests cover:
      1) Three non-empty sorted lists
      2) Mix of empty and non-empty lists
      3) All lists empty
      4) Single list
      5) Zero lists (empty input array)

    Note:
      - The merged lists reuse nodes from the original lists.
      - This test harness does not explicitly free memory (JS GC handles it).
*/

console.log("=== Test 1: three sorted lists ===");
const l1 = buildList([1, 4, 7]);
const l2 = buildList([2, 5, 8]);
const l3 = buildList([3, 6, 9]);

printList("List 1: ", l1);
printList("List 2: ", l2);
printList("List 3: ", l3);

let merged1 = mergeKLists([l1, l2, l3]);
printList("Merged: ", merged1);
console.log("(Expected: 1 2 3 4 5 6 7 8 9)\n");

console.log("=== Test 2: mix of empty and non-empty lists ===");
const a = buildList([1, 3, 5]);
const b = null;
const c = buildList([2, 4, 6, 8]);
const d = null;

printList("List A: ", a);
printList("List B: ", b);
printList("List C: ", c);
printList("List D: ", d);

let merged2 = mergeKLists([a, b, c, d]);
printList("Merged: ", merged2);
console.log("(Expected: 1 2 3 4 5 6 8)\n");

console.log("=== Test 3: all empty lists ===");
let merged3 = mergeKLists([null, null, null]);
printList("Merged: ", merged3);
console.log("(Expected: <empty>)\n");

console.log("=== Test 4: single list ===");
const single = buildList([10, 20, 30]);
printList("Input:  ", single);

let merged4 = mergeKLists([single]);
printList("Merged: ", merged4);
console.log("(Expected: 10 20 30)\n");

console.log("=== Test 5: zero lists ===");
let merged5 = mergeKLists([]);
printList("Merged: ", merged5);
console.log("(Expected: <empty>)\n");

File: code_samples/section6/problems/problem6_4/problem6_4.c
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>

/*
    This file implements "merge k sorted linked lists" using a custom
    min-heap of ListNode* (ordered by node->val).

    High-level approach:
      - Maintain a min-heap containing the current head node of each list.
      - Repeatedly extract the smallest node from the heap and append it to the
        merged result.
      - If the extracted node has a next pointer, push node->next into the heap.
      - Continue until the heap is empty.

    Why a min-heap?
      - Each extraction gives the next smallest value among the current k "front"
        nodes, making the merge efficient.

    Complexity:
      - Let N be the total number of nodes across all lists.
      - Heap size is at most k.
      - Each push/pop is O(log k).
      - Total time: O(N log k), extra space: O(k) for the heap array.

    Memory model:
      - mergeKLists() REUSES the original nodes; it does not allocate new nodes.
      - The test harness allocates the lists and frees the merged list afterward.
*/

/* ====== ListNode definition ====== */

/*
    Singly linked list node.
*/
typedef struct ListNode {
    int val;
    struct ListNode *next;
} ListNode;

/* ====== NodeMinHeap definition ====== */

/*
    NodeMinHeap stores pointers to ListNode.
    Ordering is by node->val (min-heap: smallest value at index 0).
*/
typedef struct {
    ListNode **data;  // backing array of ListNode* in heap order
    int size;         // number of elements currently in heap
    int capacity;     // allocated capacity of data[]
} NodeMinHeap;

/* ====== Heap helpers ====== */

/*
    node_heap_swap
    --------------
    Swaps two ListNode* entries in the heap array.
*/
static void node_heap_swap(ListNode **a, ListNode **b) {
    ListNode *tmp = *a;
    *a = *b;
    *b = tmp;
}

/*
    node_heap_ensure_capacity
    -------------------------
    Grows heap backing storage when size reaches capacity.

    Strategy:
      - If size < capacity: no growth needed.
      - Else: double capacity (or set to 1 if current capacity is 0).
      - realloc the pointer array (NOT the nodes themselves).

    Safety:
      - Exits on allocation failure.
*/
static void node_heap_ensure_capacity(NodeMinHeap *h) {
    if (h->size < h->capacity) return;

    int newCap = (h->capacity > 0) ? h->capacity * 2 : 1;

    ListNode **nd = (ListNode **)realloc(h->data, newCap * sizeof(ListNode *));
    if (!nd) {
        fprintf(stderr, "Out of memory in node_heap_ensure_capacity\n");
        exit(EXIT_FAILURE);
    }

    h->data = nd;
    h->capacity = newCap;
}

/*
    node_heap_bubble_up
    -------------------
    Restores min-heap property after insertion.

    Min-heap invariant:
      - Every parent node has value <= its children.

    When we insert, we append the new item at the end (index i),
    then swap it upward while it's smaller than its parent.

    Complexity: O(log n) where n = heap size.
*/
static void node_heap_bubble_up(NodeMinHeap *h, int i) {
    while (i > 0) {
        int p = (i - 1) / 2; // parent index

        // If child >= parent, heap property is satisfied.
        if (h->data[i]->val >= h->data[p]->val) break;

        // Otherwise swap and continue upward.
        node_heap_swap(&h->data[i], &h->data[p]);
        i = p;
    }
}

/*
    node_heap_heapify_down
    ----------------------
    Restores min-heap property after removing the root.

    Pop operation typically:
      - Replace root with last element
      - Decrease size
      - Push the new root down by swapping with the smaller child until valid.

    Complexity: O(log n) where n = heap size.
*/
static void node_heap_heapify_down(NodeMinHeap *h, int i) {
    int n = h->size;

    for (;;) {
        int smallest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        // Choose the smallest among i and its children (if children exist).
        if (left < n && h->data[left]->val < h->data[smallest]->val) {
            smallest = left;
        }
        if (right < n && h->data[right]->val < h->data[smallest]->val) {
            smallest = right;
        }

        // If i is already the smallest, we're done.
        if (smallest == i) break;

        // Swap downward and continue from the child index.
        node_heap_swap(&h->data[i], &h->data[smallest]);
        i = smallest;
    }
}

/* ====== Public NodeMinHeap API ====== */

/*
    node_heap_init
    --------------
    Initializes a NodeMinHeap with a given initial capacity.

    Parameters:
      - initialCapacity: number of slots to pre-allocate for ListNode* pointers.
        If <= 0, defaults to 1.
*/
void node_heap_init(NodeMinHeap *h, int initialCapacity) {
    h->size = 0;
    h->capacity = (initialCapacity > 0) ? initialCapacity : 1;

    h->data = (ListNode **)malloc(h->capacity * sizeof(ListNode *));
    if (!h->data) {
        fprintf(stderr, "Out of memory in node_heap_init\n");
        exit(EXIT_FAILURE);
    }
}

/*
    node_heap_free
    --------------
    Frees the heap's backing array (not the ListNode objects!).
*/
void node_heap_free(NodeMinHeap *h) {
    free(h->data);
    h->data = NULL;
    h->size = 0;
    h->capacity = 0;
}

/*
    node_heap_push
    --------------
    Inserts a ListNode* into the heap (by node->val ordering).

    Important:
      - Caller must not pass NULL here; bubble_up compares node->val.
*/
void node_heap_push(NodeMinHeap *h, ListNode *node) {
    node_heap_ensure_capacity(h);
    h->data[h->size] = node;
    node_heap_bubble_up(h, h->size);
    h->size++;
}

/*
    node_heap_peek
    --------------
    Returns the smallest node (root) without removing it, or NULL if empty.
*/
ListNode *node_heap_peek(const NodeMinHeap *h) {
    if (h->size == 0) return NULL;
    return h->data[0];
}

/*
    node_heap_pop
    -------------
    Removes and returns the smallest node (root), or NULL if empty.

    Steps:
      - Save root
      - Decrease size
      - Move last item to root (if any remain)
      - Heapify down

    Note:
      - The returned node is still part of the original lists; we do not free it.
*/
ListNode *node_heap_pop(NodeMinHeap *h) {
    if (h->size == 0) return NULL;

    ListNode *root = h->data[0];
    h->size--;

    if (h->size > 0) {
        h->data[0] = h->data[h->size];
        node_heap_heapify_down(h, 0);
    }

    return root;
}

/* ====== mergeKLists implementation ====== */

/*
    mergeKLists
    -----------
    Merges k sorted singly linked lists into one sorted list.

    Parameters:
      - lists: array of ListNode* heads (may include NULLs)
      - k: number of lists in the array

    Returns:
      - head pointer to the merged sorted list, or NULL if all lists are empty.

    Implementation details:
      - Push the head of each non-empty list into the min-heap.
      - Repeatedly pop the smallest node and append it to the output list.
      - When popping a node, push its next node (if any) into the heap.

    Memory:
      - Uses a stack-allocated dummy node to simplify output list construction.
      - Reuses existing nodes (does not allocate new nodes).
*/
ListNode *mergeKLists(ListNode **lists, int k) {
    // Initialize heap with capacity k: at most one node per list is in the heap at once.
    NodeMinHeap h;
    node_heap_init(&h, k);

    // Seed heap with the head of each non-empty list.
    for (int i = 0; i < k; i++) {
        if (lists[i] != NULL) {
            node_heap_push(&h, lists[i]);
        }
    }

    // Dummy head simplifies appending: output starts at dummy.next
    ListNode dummy = {0, NULL};
    ListNode *tail = &dummy;

    // Keep extracting smallest node and pushing its successor.
    while (h.size > 0) {
        ListNode *node = node_heap_pop(&h);

        // Append node to merged list
        tail->next = node;
        tail = tail->next;

        // If this node came from a list that still has nodes, push the next one.
        if (node->next != NULL) {
            node_heap_push(&h, node->next);
        }
    }

    // IMPORTANT:
    // Ensure the merged list terminates (usually already true, but safe if inputs
    // had unexpected sharing/cycles).
    tail->next = NULL;

    node_heap_free(&h);
    return dummy.next;
}

/* ====== Test helpers ====== */

/*
    create_list_from_array
    ----------------------
    Allocates a new linked list whose values match arr[0..n-1].

    Returns:
      - head pointer (or NULL if n == 0).

    Ownership:
      - Caller owns the returned list and must free it (free_list),
        unless the nodes are reused/merged and freed elsewhere.
*/
static ListNode *create_list_from_array(const int *arr, int n) {
    ListNode *head = NULL;
    ListNode *tail = NULL;

    for (int i = 0; i < n; i++) {
        ListNode *node = (ListNode *)malloc(sizeof(ListNode));
        if (!node) {
            fprintf(stderr, "Out of memory in create_list_from_array\n");
            exit(EXIT_FAILURE);
        }

        node->val = arr[i];
        node->next = NULL;

        if (!head) {
            // First node becomes head and tail
            head = tail = node;
        } else {
            // Append to the tail
            tail->next = node;
            tail = node;
        }
    }

    return head;
}

/*
    print_list
    ----------
    Prints a linked list as space-separated values, preceded by a label.

    If head is NULL, it prints just the label and a newline.
*/
static void print_list(const char *label, ListNode *head) {
    printf("%s", label);
    for (ListNode *cur = head; cur != NULL; cur = cur->next) {
        printf("%d ", cur->val);
    }
    printf("\n");
}

/*
    free_list
    ---------
    Frees all nodes in a singly linked list.

    Assumes:
      - Nodes were allocated with malloc.
      - List is acyclic.

    Important:
      - In these tests, mergeKLists reuses the original nodes,
        so freeing the merged list frees the original lists too.
*/
static void free_list(ListNode *head) {
    while (head) {
        ListNode *next = head->next;
        free(head);
        head = next;
    }
}

/* ====== Test function ====== */

/*
    test_mergeKLists
    ----------------
    Runs a few basic tests of mergeKLists:

      Test 1: 3 non-empty sorted lists
      Test 2: mixture of empty and non-empty lists
      Test 3: all lists empty

    Each test prints input lists and merged output.
*/
void test_mergeKLists(void) {
    // Example 1: three non-empty sorted lists
    int a1[] = {1, 4, 7};
    int a2[] = {2, 5, 8};
    int a3[] = {3, 6, 9};

    ListNode *l1 = create_list_from_array(a1, (int)(sizeof(a1)/sizeof(a1[0])));
    ListNode *l2 = create_list_from_array(a2, (int)(sizeof(a2)/sizeof(a2[0])));
    ListNode *l3 = create_list_from_array(a3, (int)(sizeof(a3)/sizeof(a3[0])));

    ListNode *lists1[] = { l1, l2, l3 };
    int k1 = (int)(sizeof(lists1)/sizeof(lists1[0]));

    printf("=== Test 1: 3 sorted non-empty lists ===\n");
    print_list("List 1: ", l1);
    print_list("List 2: ", l2);
    print_list("List 3: ", l3);

    ListNode *merged1 = mergeKLists(lists1, k1);
    print_list("Merged: ", merged1);
    printf("(Expected: 1 2 3 4 5 6 7 8 9)\n\n");

    // merged list reuses original nodes, so freeing merged1 frees l1/l2/l3 too
    free_list(merged1);

    // Example 2: some empty lists
    int b1[] = {1, 3, 5};
    int b3[] = {2, 4, 6, 8};

    ListNode *m1 = create_list_from_array(b1, (int)(sizeof(b1)/sizeof(b1[0])));
    ListNode *m2 = NULL;
    ListNode *m3 = create_list_from_array(b3, (int)(sizeof(b3)/sizeof(b3[0])));
    ListNode *m4 = NULL;

    ListNode *lists2[] = { m1, m2, m3, m4 };
    int k2 = (int)(sizeof(lists2)/sizeof(lists2[0]));

    printf("=== Test 2: mix of empty and non-empty lists ===\n");
    print_list("List 1: ", m1);
    print_list("List 2: ", m2);
    print_list("List 3: ", m3);
    print_list("List 4: ", m4);

    ListNode *merged2 = mergeKLists(lists2, k2);
    print_list("Merged: ", merged2);
    printf("(Expected: 1 2 3 4 5 6 8)\n\n");

    free_list(merged2);

    // Example 3: all empty
    ListNode *lists3[] = { NULL, NULL };
    int k3 = (int)(sizeof(lists3)/sizeof(lists3[0]));

    printf("=== Test 3: all lists empty ===\n");
    ListNode *merged3 = mergeKLists(lists3, k3);
    print_list("Merged: ", merged3);
    printf("(Expected: <empty line>)\n\n");

    // merged3 is NULL, nothing to free
    (void)merged3;
    (void)k3; // k3 is computed for completeness; not otherwise used
}

/* ====== main ====== */

int main(void) {
    test_mergeKLists();
    return 0;
}
